### PR TITLE
Switch all libraries to the 2021 edition

### LINKED ISSUE #88638: Switch library to Rust 2021
This is blocked on the beta/bootstrap compiler being updated to 1.56.

### PR DESCRIPTION
**Pull request by fee1-dead**
The fix for https://github.com/rust-lang/rust/issues/88638#issuecomment-996620107 is to simply add const-stability for these functions.

r? @m-ou-se 

Closes #88638.

### CODE REVIEW

**Review by jyn514 — APPROVED** [review]

**Comment by m-ou-se** [comment]
Thanks!

@bors r+

**Comment by m-ou-se** [comment]
> Very large regression in instruction counts (up to 10.3% on `full` builds of `issue-88862`)

Oh, that's quite significant. We should definitely investigate that before merging this.

@bors r-

**Comment by jyn514** [comment]
@m-ou-se perf has been super sporadic lately, even changes to rustdoc have had wildly different perf results on check/opt builds. I don't think this change is related.

File: src/test/mir-opt/issues/issue_59352.num_to_digit.PreCodegen.after.mir
Comment (eddyb) [inline src/test/mir-opt/issues/issue_59352.num_to_digit.PreCodegen.after.mir:22]: `Option::unwrap` not MIR-inlining anymore suggests its body is more complex (maybe not enough to stop LLVM from inlining itself, but at least making it more work to simplify).

If a lot of `panic!` users in the standard library suddenly increased in complexity, this could *easily* explain the regressions, on their own (though without alleviating them, I can't be sure there aren't other causes too).

In general, if something is causing non-trivial changes in MIR tests, a PR shouldn't be just `--bless`'d and merged, but an additional review from @rust-lang/wg-mir-opt (or anyone else in @rust-lang/compiler with MIR experience) should be required.

Even if some of these MIR optimizations are not on by default, they're good canaries for potential issues down the line.

**Comment by eddyb** [comment]
> @m-ou-se perf has been super sporadic lately, even changes to rustdoc have had wildly different perf results on check/opt builds. I don't think this change is related.

Keep in mind this has a much larger semantic impact, which I could maybe dismiss if not for the MIR tests (and some other weirder tests) clearly showing regressions in generated AST/MIR.

<hr/>

Also, doesn't this PR change user-observable behavior, if macros like `unreachable!()` and `todo!()` switch to a different `panic!` implementation? [As you can see on (`edition=2021`) playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4a61e6303c6b52899fffed6f2054d596):
```rust
fn main() {
    unreachable!("foo {}");
}
```
prints:
```
thread 'main' panicked at 'internal error: entered unreachable code: foo {}', src/main.rs:2:5
```

Sadly I haven't paid enough attention to know for sure if *all* the macros that use `panic!` were supposed to be changed, but this doesn't produce the same compiler error as `panic!("foo {}")`.

**Comment by m-ou-se** [comment]
> Also, doesn't this PR change user-observable behavior, if macros like `unreachable!()` switch to a different `panic!` implementation?

Looks like `unreachable!()` is implemented in a way that does depend on the edition, unlike `todo!()` and `unimplemented!()`. I must've missed that. I'll open an issue for that, and check how commonly it's passed any arguments 'in the wild'.

File: src/test/mir-opt/issues/issue_59352.num_to_digit.PreCodegen.after.mir
Comment (m-ou-se) [reply src/test/mir-opt/issues/issue_59352.num_to_digit.PreCodegen.after.mir:22]: Checking now what happens when we implement `unwrap()` with `panic()` (the function) instead of `panic!()` (the macro).

Comment (m-ou-se) [reply src/test/mir-opt/issues/issue_59352.num_to_digit.PreCodegen.after.mir:22]: https://github.com/rust-lang/rust/pull/92140

**Comment by m-ou-se** [comment]
> Also, doesn't this PR change user-observable behavior, if macros like `unreachable!()` switch to a different `panic!` implementation?

I took a closer look. This PR won't change anything there. It's *not* the case that unreachable!() is dependent on the edition of core rather than the user's crate. Instead, it just has a broken implementation, independent of the edition of anything. It's definitely a bug, but unrelated to this PR.

**Comment by m-ou-se** [comment]
After https://github.com/rust-lang/rust/pull/92155 is merged, this PR should be ready to go.

**Comment by m-ou-se** [comment]
@fee1-dead #92155 is merged. Can you rebase this PR?

**Comment by eddyb** [comment]
I happened to have https://github.com/rust-lang/rust/pull/92140#issuecomment-998375613 open in a tab next to this one, and this part is identical between the two runs:
> - Large regression in instruction counts (up to 3.2% on `incr-patched: b9b3e592dd cherry picked` builds of `style-servo`)

So I suspect at least that one is not spurious. Also, both it and the smaller regressions seem to be mostly in LLVM so it could be a bit of extra codegen work, but at least it seems to affect fewer benchmarks.

I haven't looked closer at any of these, and honestly I wish we had a good "CGU diff" mechanism (cc @wesleywiser) to be able to tell between "more (instances of) functions are getting codegen'd" and "there's more time spent for existing functions" (and "which functions have the worst regression" would be the dream).

Overall, these few regressions may be easier to overlook, but I'm in no position to decide that myself.

**Comment by m-ou-se** [comment]
Reassigning to Mark, who hopefully has some ideas on what to do with these perf results.

r? @Mark-Simulacrum

**Comment by Mark-Simulacrum** [comment]
Not entirely sure on the causes of the style-servo regression, though I suspect it's the same root cause. For the tokio-webpush-simple regression, I'm fairly confident in saying that the root cause is that functions in the standard libraries (e.g., `core::sync::atomic::atomic_load`) are now slightly more expensive to compile and at runtime, which means that end-users see worse performance (compile and presumably slight impact on runtime) after a migration of core/std from 2018 to 2021. I suspect the right next path here is to (1) land this PR, it's not worth blocking on this, and (2) work to bring 2021 panic to parity codegen-wise with 2015/2018. That may come from more general fmt::Arguments improvements, which would likely be more broadly desirable.

@m-ou-se want to get another set of eyes on this in case you disagree with that plan and/or know of an existing issue tracking the now-worse codegen here.

```rust
pub fn foo() {
    panic!("test");
}

pub fn bar() {
    panic!("bar");
}
```

compiles to:

Edition 2015/2018:

```asm
example::foo:
        push    rax
        lea     rdi, [rip + .L__unnamed_1]
        lea     rdx, [rip + .L__unnamed_2]
        mov     esi, 4
        call    std::panicking::begin_panic
        ud2

example::bar:
        push    rax
        lea     rdi, [rip + .L__unnamed_3]
        lea     rdx, [rip + .L__unnamed_4]
        mov     esi, 3
        call    std::panicking::begin_panic
        ud2
```

Edition 2021:

```asm
example::foo:
        sub     rsp, 56
        lea     rax, [rip + .L__unnamed_1]
        mov     qword ptr [rsp + 8], rax
        mov     qword ptr [rsp + 16], 1
        mov     qword ptr [rsp + 24], 0
        lea     rax, [rip + .L__unnamed_2]
        mov     qword ptr [rsp + 40], rax
        mov     qword ptr [rsp + 48], 0
        lea     rsi, [rip + .L__unnamed_3]
        lea     rdi, [rsp + 8]
        call    qword ptr [rip + core::panicking::panic_fmt@GOTPCREL]
        ud2

example::bar:
        sub     rsp, 56
        lea     rax, [rip + .L__unnamed_4]
        mov     qword ptr [rsp + 8], rax
        mov     qword ptr [rsp + 16], 1
        mov     qword ptr [rsp + 24], 0
        lea     rax, [rip + .L__unnamed_2]
        mov     qword ptr [rsp + 40], rax
        mov     qword ptr [rsp + 48], 0
        lea     rsi, [rip + .L__unnamed_5]
        lea     rdi, [rsp + 8]
        call    qword ptr [rip + core::panicking::panic_fmt@GOTPCREL]
        ud2
```

Godbolt link - https://rust.godbolt.org/z/jxEvb5Pfb

**Comment by m-ou-se** [comment]
Thanks!

Yes, those snippers are just `fmt::Arguments` being large and inefficient. I've been experimenting with different representations of that type for while now, but looks like it's time to actually start pushing one of those forward. We could make the trivial case of a `fmt::Arguments` have an identical representation as a `&str` (and use the high bit of the str's length to indicate the non-trivial case). There's a few other options as well. All of them result in a `fmt::Arguments` that's two pointers in size, rather than six, just like a `&str`.

> I suspect the right next path here is to (1) land this PR, it's not worth blocking on this, and (2) work to bring 2021 panic to parity codegen-wise with 2015/2018. That may come from more general fmt::Arguments improvements, which would likely be more broadly desirable.

Agreed.

@bors r+

**Comment by fee1-dead** [comment]
```
    Compiling rustc_mir_build v0.0.0 (/Users/runner/work/rust/rust/compiler/rustc_mir_build)
== clock drift check ==
src/ci/run.sh: line 115:  2638 Terminated: 15          sh -x -c "$SCRIPT"
  local time: Fri Jan  7 09:40:00 UTC 2022
dyld: Library not loaded: /usr/local/opt/libidn2/lib/libidn2.0.dylib
  Referenced from: /usr/local/opt/curl/bin/curl
  Reason: no suitable image found.  Did find:
	/usr/local/opt/libidn2/lib/libidn2.0.dylib: code signature in (/usr/local/opt/libidn2/lib/libidn2.0.dylib) not valid for use in process using Library Validation: system is shutting down
	/usr/local/lib/libidn2.0.dylib: code signature in (/usr/local/lib/libidn2.0.dylib) not valid for use in process using Library Validation: system is shutting down
	/usr/local/Cellar/libidn2/2.3.2/lib/libidn2.0.dylib: code signature in (/usr/local/Cellar/libidn2/2.3.2/lib/libidn2.0.dylib) not valid for use in process using Library Validation: system is shutting down
  network time: == end clock drift check ==
```

:(

**Comment by Mark-Simulacrum** [comment]
Per https://github.com/rust-lang/rust/pull/92068#issuecomment-1005229569, marking regression as triaged.

**Pull request merged**
